src/System.Net.Primitives/src/System/Net/HostInformationPal.Windows.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Net.NetworkInformation
+{
+    internal class HostInformationPal
+    {
+        // Changing this information requires a reboot, so it's safe to cache.
+        private static Interop.IpHlpApi.FIXED_INFO s_fixedInfo;
+        private static bool s_fixedInfoInitialized;
+        private static object s_syncObject = new object();
+
+        private static Interop.IpHlpApi.FIXED_INFO GetFixedInfo()
+        {
+            uint size = 0;
+            SafeLocalAllocHandle buffer = null;
+            Interop.IpHlpApi.FIXED_INFO fixedInfo = new Interop.IpHlpApi.FIXED_INFO();
+
+            // First we need to get the size of the buffer
+            uint result = Interop.IpHlpApi.GetNetworkParams(SafeLocalAllocHandle.InvalidHandle, ref size);
+
+            while (result == Interop.IpHlpApi.ERROR_BUFFER_OVERFLOW)
+            {
+                try
+                {
+                    // Now we allocate the buffer and read the network parameters.
+                    buffer = Interop.mincore_obsolete.LocalAlloc(0, (UIntPtr)size);
+                    if (buffer.IsInvalid)
+                    {
+                        throw new OutOfMemoryException();
+                    }
+
+                    result = Interop.IpHlpApi.GetNetworkParams(buffer, ref size);
+                    if (result == Interop.IpHlpApi.ERROR_SUCCESS)
+                    {
+                        fixedInfo = Marshal.PtrToStructure<Interop.IpHlpApi.FIXED_INFO>(buffer.DangerousGetHandle());
+                    }
+                }
+                finally
+                {
+                    if (buffer != null)
+                    {
+                        buffer.Dispose();
+                    }
+                }
+            }
+
+            // If the result include there being no information, we'll still throw
+            if (result != Interop.IpHlpApi.ERROR_SUCCESS)
+            {
+                throw new NetworkInformationException((int)result);
+            }
+            return fixedInfo;
+        }
+
+        private static void EnsureFixedInfo()
+        {
+            if (!s_fixedInfoInitialized)
+            {
+                lock (s_syncObject)
+                {
+                    if (!s_fixedInfoInitialized)
+                    {
+                        s_fixedInfo = GetFixedInfo();
+                        s_fixedInfoInitialized = true;
+                    }
+                }
+            }
+        }
+
+        public static string GetHostName()
+        {
+            EnsureFixedInfo();
```

Comment:
@pgavlin Looking at currently used code I see two changes from the original HostInformation.cs:
1. The internal contract changed to have methods instead of properties. I think that's ok but I'm curious what was the main reason.
2. You have extracted the EnsureFixedInfo() to remove duplicate code. I think this is fine although I believe that the correct singleton pattern requires _volatile_ for `s_fixedInfoInitialized`. (This was missing from the original implementation as well.)

@stephentoub I usually stay away from static initialization that calls into external APIs and may throw. It's very tricky to get right: once the initializer throws, that type can't ever be used throughout the lifetime of the AppDomain. With the current implementation, the application developer will at least have a way to work-around any such issue until we make a fix and he's ready to take a new version of our implementation.